### ISSUE TITLE
Using "\$0" in iemguis behave differently

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#560]
We now can use backslashes in Pd, and I just realize this can help us dealing with "$0" in messages. Using something like "\\$0-a" can set a literal "$0-a" symbol. One use case is to set a receive name in a GUI, and it gets saved as a literal "$0" and not an expanded version.

On the other hand, I'd also expect something like 

~~~
;
\$0-a bang
~~~

to work, but it doesn't. Any special reason why this should not work?

See test patch attached

[test.pd.zip](https://github.com/pure-data/pure-data/files/2872300/test.pd.zip)

**Comment by porres** [comment]
I now realise this doesn't work by setting a [send] object either. So it works fine for iemguis, but not other objects in Pd. Here's another test patch and a screenshot. 

[test-2.pd.zip](https://github.com/pure-data/pure-data/files/3139179/test-2.pd.zip)

<img width="402" alt="Screen Shot 2019-05-02 at 14 24 58" src="https://user-images.githubusercontent.com/5360142/57094113-14dfcd00-6ce6-11e9-8144-71d19e7081a6.png">

**Comment by porres** [comment]
I'm more curious to understand why this happens. For the use case, this helps dealing with '$0' in messages, but for that, what I really hope and can't wait for is this ;) => https://github.com/pure-data/pure-data/pull/347

**Comment by Spacechild1** [comment]
I think this is rather a bug in the iemguis. \$ is meant to escape the dollarsym so it doesn't get expanded. 
This means "\$0" is a symbol and won't expand to the $0 value in a typed message. But somehow the iemguis expand the \$0 to $0 nevertheless... Can you change the title/description or open a new issue?

**Comment by porres** [comment]
how would you like to call it? :)

**Comment by Spacechild1** [comment]
maybe something like "backslash escaping doesn't work in iemguis". seems like \ characters are just dropped.

**Comment by porres** [comment]
> This means "\$0" is a symbol and won't expand to the $0 value in a typed message. But somehow the iemguis expand "\$0" to $0 nevertheless... Can you change the title/description or open a new issue?

Actually, I don't think they get expanded. In fact it's treated as a symbol, which just works as well...

**Comment by Spacechild1** [comment]
since `[bang( ->  [s $0-a]` "works", this means `\\$0` in the iemgui receive name gets expanded to the $0 value. what's actually happening I think is that backslashes are simply dropped. haven't looked at the code though.

**Comment by porres** [comment]
Well, what's happening here, and probably it's what should happen anyway, is that a 'symbol \$0-a' message sent to a [send] object or used inside a message communicates with a [r \$0-a] object, and the '$0' does not get expanded - which means if I open this same patch twice, then the send/receive are global, not local.
 
<img width="273" alt="Screen Shot 2019-05-03 at 00 56 56" src="https://user-images.githubusercontent.com/5360142/57119189-98c5a380-6d3e-11e9-8b3c-be47b9bd6d7b.png">

**Comment by porres** [comment]
> what's actually happening I think is that backslashes are simply dropped. 

yup, definitely iemguis deals with something like that, unlike [send]

**Comment by porres** [comment]
by the way, it's not the first time iemguis seem to deal with this in a different way :) => see https://github.com/pure-data/pure-data/issues/558

**Comment by porres** [comment]
Had a quick look in the code, and iemguis use the same 'canvas_realizedollar' function that atom boxes uses...

**Comment by Spacechild1** [comment]
ok, turns out the behavior of the iemguis is totally expected:

when you send `[receive \$0-a(`, what you're actually asking is that the` $0` doesn't get expanded *in the message box* but you do want the GUI object to expand it. BTW, in the iemguis `[receive \$0-a(` behaves the same as `[receive #0-a(`.

but now here comes the real bug: I would have guessed that setting the receive name to a literal, unexpanded dollsym would go something like this: `[receive \\\$0-a(`. saving the patch shows that the receive name is correctly saved in its unexpanded version but `[s \$0-a]` doesn't work.

I think the problem is that `canvas_realizedollar`/`binbuf_realizedollsym` assume that a symbol is a dollarsymbol once it contains at least one $ character and then ask `binbuf_expanddollsy` to expand it. none of these functions take backslashes into account. you can check this by replacing the $0 in `[s \$0-a]` with `[s \\1023-a]` (where 1023 is the current $0 value) - this will work!

maybe the tests for $ in `binbuf_realizedollsym` should be modified to handle backslashes. what does @umlaeute think? OTOH, why would someone want to use a literal `$0` in a send/receive name!?

also, typing `\$0-a` in the GUI properties doesn't work as well but the behavior differs between iemguis and atoms: the iemgui strips the backslash *including* the dollarsym (leaving only` 0-a`) while the atoms just ignore the backslash.

can you change the bug description again? :-D something like "escaping $ characters doesn't work in GUI send/receive names", but again: why would anyone want to do this?

**Comment by porres** [comment]
> but now here comes the real bug: I would have guessed that setting the receive name to a literal, "unexpandable" dollsym would go something like this: [receive \\\$0-a(. 

Sorry, you completely lost me here. So, do you think that [receive \$0-a( shouldn't work, but [receive \\\$0-a( should instead? If so, why?

> looking at the saved patch shows that the receive name is correctly saved as \\\$0-a but [s \$0-a] doesn't work.

Well, here's what I tested, and it worked. I saved, reopened the patch, and sending the message gets received by bang. Check the text patch as well.

<img width="595" alt="Screen Shot 2019-05-03 at 14 25 22" src="https://user-images.githubusercontent.com/5360142/57154539-82a3fb80-6daf-11e9-8dc8-ae37b3e35f2a.png">

Anyway, I gotta say I'm confused with all the extra '\\\\\\\\\' and stuff. And why the patch there, for instance, turns |receive \\\$0-bng( into "#X msg 103 103 receive \\\\\\\$0-bng;" there are 3 backlashes in the message box but SEVEN in the saved file.

I really need to better understand what's the deal with '\' in Pd, how it handles it and everything. I hope you can help me :) 

thanks

**Comment by Spacechild1** [comment]
when you quote code blocks containing `\` characters, make sure to preserve the back ticks, otherwise some `\` characters get lost, which doesn't really help to clear the confusion :-). the quotes above should really read:

>but now here comes the real bug: I would have guessed that setting the receive name to a literal, "unexpandable" dollsym would go something like this:` [receive \\\$0-a(`.

> looking at the saved patch shows that the receive name is correctly saved as `\\\$0-a` but `[s \$0-a]` doesn't work.

**Comment by Spacechild1** [comment]
> Sorry, you completely lost me here. So, do you think that [receive $0-a( shouldn't work, but [receive \\$0-a( should instead? If so, why?

did you miss some backslashes here? anyway, both should work, but they are different. `[receive \\\$0-a(` should set the receive label to an "unexpandable" symbol `\$0-a` - where the dollar character is escaped by the backslash - while `[receive \$0-a(` sets it to `$0-a` which the GUI objects expands to the canvas $0 value.

**Comment by Spacechild1** [comment]
> Well, here's what I tested, and it worked. I saved, reopened the patch, and sending the message gets received by bang. Check the text patch as well.

you're right! the reason it works after reopening is that `\$0-a` (actually saved as `\\\$0-a`) is evaluated by `binbuf_eval` - which correctly handles backslashes. if you just use [receive(, the symbol is "evaluated" by `canvas_realizedollar` - which doesn't handle backslashes.

**Comment by porres** [comment]
Ok, so for some reason I still don't know, both Pd and github's comments have a way of disappearing with backlashes :) - I hope you can help me understand how this works. I can get that '$0' in patches get saved with a prepended backlash, this is a way to avoid '$0' getting expanded and all. Now, if you have a message with a backlash, why does it add 2 more backlashes? I can see why you should also add one backlash, but why two?

**Comment by porres** [comment]
Perhaps the answer to my question may also explain this https://github.com/pure-data/pure-data/issues/615

**Comment by Spacechild1** [comment]
if you want to type `\` characters, a) use code formatting or b) escape them with another \\

> Anyway, I gotta say I'm confused with all the extra '\\\\' and stuff.

the reason is that certain characters have a special meaning in Pd, notable the semicolon, comma, dollarsign and backslash (the escape character). when we want to type those symbols as literal characters, we have to escape them with a backslash. in the same way, Pd has to escape them when saving them as text. note that a literal backslash is escaped by a backslash. now back to your example. let's do it in ascending complexity:

`1)` `#X msg 103 103 receive $0-bng;`
the $0 is immediately expanded to the $0 value, so we might see `[receive 1003-bng(`. this isn't very useful, the expansion should happen when passing the args to the object - not in the text itself - that's why Pd saves it like this:
`2)` `#X msg 103 103 receive \$0-bng;`
we get `[receive $0-bng(`. 

now let's say we want a literal $. for this we need to make sure that a backslash appears before the $. we can type it, but when Pd saves the patch, it must escape both the backslash and the $0:
`3)` `#X msg 103 103 receive \\\$0-bng;`
we get `[receive \$0-bng(`.

OK, when clicking the message box, the message is evaluated to an anything message with the selector `receive` and a symbol with the string `$0-bng`. the GUI object will evaluate this symbol and expand `$0-bng` to something like `1003-bng` but also keep an unexpanded version (which is saved as `\$0-bng`). this is probably what most people want. 

but let's say I don't want the GUI object to expand the $0 part. this means I must send a symbol with the string `\$0-bng` (not the backslash!), so I have to escape *both* the backslash and the dollarsym in the message box: `[receive \\\$-bng(`. the first backslash escapes the second one, and the third backslash escapes the $. Now when Pd saves this message box, it has to escape all those special characters, like in the examples above and you end up with 7 backslashes in a row:
`4)` `#X msg 103 103 receive \\\\\\\$0-bng;`

I know it gets absurd but I hope you can see the pattern!

**Comment by Spacechild1** [comment]
TL;DR: here's the pattern - with spaces added for legibility:
`$` -> `\$`
`\` -> `\\`
`\ $` -> `\\ \$`
`\ \ \ $` -> `\\ \\ \\ \$`

**Comment by porres** [comment]
thanks

**Comment by porres** [comment]
Well, anyway, I do agree the behaviour in iemgui is totally expected. In fact, I opened this issue because I thought we should have the same behaviour when using it inside messages. But now that I think of it, no it should not.

Things should be as they are, with iemguis and maybe other GUIs (like the ones I'm writing) being able to get a literal $0 and expand it, but also save it unexpanded. Since this is not the issue with something like [send] (meaning, send does not get saved with an internal state) - this is not needed!

Now, I guess I can close this issue... but you seem to have found other issues that I already forget. So maybe you open a new one than I close this?

**Comment by Spacechild1** [comment]
> So maybe you open a new one than I close this?

ok!

**Comment by porres** [comment]
and what is the issue anyway, that canvas_realizedollar doesn't handle backslashes?

**Comment by porres** [comment]
> maybe the tests for $ in binbuf_realizedollsym should be modified to handle backslashes. what does @umlaeute think? OTOH, why would someone want to use a literal $0 in a send/receive name!?

For no reason. For one, I was trying to do something else...

**Comment by Spacechild1** [comment]
I think you can close this issue in favor of https://github.com/pure-data/pure-data/issues/558.

**Comment by porres** [comment]
seems unrelated but fine :)

**Comment by porres** [comment]
this issue seems more related https://github.com/pure-data/pure-data/issues/622
